server/datastore/mysql/hosts.go

```diff
@@ -1913,44 +1923,59 @@ WHERE
 		return nil, err
 	}
 
-	osVersions := &fleet.OSVersions{
+	res := &fleet.OSVersions{
 		CountsUpdatedAt: row.UpdatedAt,
+		OSVersions:      []fleet.OSVersion{},
 	}
 
+	var counts []fleet.OSVersion
 	if row.JSONValue != nil {
-		err := json.Unmarshal(*row.JSONValue, &osVersions.OSVersions)
-		if err != nil {
-			return nil, err
+		if err := json.Unmarshal(*row.JSONValue, &counts); err != nil {
+			return nil, ctxerr.Wrap(ctx, err)
 		}
 	}
 
-	// filter by os versions by platform
+	// filter counts by platform
 	if platform != nil {
 		var filtered []fleet.OSVersion
-		for _, osVersion := range osVersions.OSVersions {
-			if *platform == osVersion.Platform {
-				filtered = append(filtered, osVersion)
+		for _, os := range counts {
+			if *platform == os.Platform {
+				filtered = append(filtered, os)
 			}
 		}
-		osVersions.OSVersions = filtered
+		counts = filtered
 	}
 
-	// filter by os versions by os id
-	if osID != nil {
-		var filtered []fleet.OSVersion
-		for _, osVersion := range osVersions.OSVersions {
-			if *osID == osVersion.ID {
-				filtered = append(filtered, osVersion)
-			}
+	// aggregate counts by name and version
+	byNameVers := make(map[string]fleet.OSVersion)
+	for _, os := range counts {
+		key := fmt.Sprintf("%s %s", os.NameOnly, os.Version)
```

Comment:
Nit: I would move the filtering logic inside the loop to avoid doing un-necessary allocations:
```
for _, os := range counts {
  if name != nil && 
     version != nil && 
     *name != os.NameOnly && 
     *version != os.Version {
    continue
  }
....
```